setup_gui.py

```diff
@@ -0,0 +1,294 @@
+        self.readback_label.showUnits = True
+        
+        # Putting this here because it otherwise gets garbage collected (?!)
+        self.spinbox: QDoubleSpinBox = QDoubleSpinBox()
+        self.spinbox.setToolTip("Press enter to update CM and Linac spinboxes")
+        self.spinbox.setValue(5)
+        
+        while caget(self.amax_pv.pvname) is None:
+            print(f"waiting for {self.amax_pv.pvname} to connect")
+            sleep(0.0001)
+        
+        self.spinbox.setRange(0, caget(self.amax_pv.pvname))
+        
+        self.status_label: QLabel = QLabel("Ready for Setup")
+    
+    def connect_spinbox(self):
+        self.spinbox.editingFinished.connect(self.cm_spinbox_update_func)
+        self.amax_pv.add_callback(self.amax_callback)
+    
+    def launch_worker(self):
+        self.worker = Worker(SETUP_CRYOMODULES[self.cm].cavities[self.number],
+                             self.spinbox.value())
+        self.worker.error.connect(print)
+        self.worker.error.connect(self.status_label.setText)
+        self.worker.error.connect(handle_error)
+        self.worker.status.connect(self.status_label.setText)
+        self.worker.status.connect(print)
+        self.worker.start()
+    
+    def amax_callback(self, value, **kwargs):
+        self.spinbox.setRange(0, value)
+        self.cm_spinbox_range_func()
+
+
+@dataclasses.dataclass
+class GUICryomodule:
+    linac_idx: int
+    name: str
+    
+    def __post_init__(self):
+        
+        self.spinbox: QDoubleSpinBox = QDoubleSpinBox()
+        self.spinbox.setValue(40)
+        self.spinbox.setToolTip("Press enter to update cavity spinboxes")
+        self.readback_label: PyDMLabel = PyDMLabel(init_channel=f"ACCL:L{self.linac_idx}B:{self.name}00:AACTMEANSUM")
+        self.setup_button: QPushButton = QPushButton(f"Set Up CM{self.name}")
+        self.setup_button.clicked.connect(self.launch_cavity_workers)
+        self.gui_cavities: Dict[int, GUICavity] = {}
+        
+        for cav_num in range(1, 9):
+            gui_cavity = GUICavity(cav_num,
+                                   f"ACCL:L{self.linac_idx}B:{self.name}{cav_num}0:",
+                                   self.name, self.update_amp, self.update_max_amp)
+            self.gui_cavities[cav_num] = gui_cavity
+        
+        self.max_amp = 0
+        self.update_max_amp()
+        
+        for gui_cavity in self.gui_cavities.values():
+            gui_cavity.connect_spinbox()
+        
+        self.spinbox.editingFinished.connect(self.set_cavity_amps)
+    
+    def update_max_amp(self):
+        self.max_amp = 0
+        
+        for gui_cavity in self.gui_cavities.values():
+            while caget(gui_cavity.amax_pv.pvname) is None:
+                print(f"waiting for {gui_cavity.amax_pv.pvname} to connect")
+                sleep(0.0001)
+            
+            self.max_amp += caget(gui_cavity.amax_pv.pvname)
+        
+        print(f"CM{self.name} max amp: {self.max_amp}")
```

Comment:
I don't think it's the print statement that takes forever lol I think it's the determination of the maximum amplitude, which I do need to do unless we make a pv that's the sum of all the amaxes in a cryomodule. It needs to be a calc PV though so that it updates with the amaxes and I don't know how to do that

Reply:
Oh, you know what? I was thinking of Matlab :) Adding a print statement to something like that will make it run a lot slower